contracts/src/interfaces/IStakeTable.sol

```diff
@@ -0,0 +1,129 @@
+    struct Node {
+        address account;
+        StakeType stakeType;
+        uint64 balance;
+        uint64 registerEpoch;
+        uint64 exitEpoch;
+        EdOnBN254.EdOnBN254Point schnorrVK;
+    }
+
+    // === Table State & Stats ===
+
+    /// @notice Get the commitment of the stake table used in current voting (i.e. snapshot at
+    /// the start of last epoch)
+    function votingStakeTableCommitment() external view returns (bytes32);
+    /// @notice Get the commitment of the stake table frozen for change (i.g. snapshot at the start
+    /// of the current epoch)
+    function frozenStakeTableCommitment() external view returns (bytes32);
+
+    /// @notice Get the total stake of the registered keys in the latest stake table (Head)
+    function totalStake() external view returns (uint256);
+    /// @notice Get the total number of stakers, uniquely identified by their `blsVK`
+    function totalKeys() external view returns (uint32);
+    /// @notice Get the total stake of the registered keys in the voting stake table
+    /// (LastEpochStart)
+    function totalVotingStake() external view returns (uint256);
+
+    /// @notice Look up the balance of `blsVK`
+    function lookup(BN254.G1Point calldata blsVK) external view returns (uint64);
+    /// @notice Look up the full `Node` state associated with `blsVK`
+    function fullLookup(BN254.G1Point calldata blsVK) external view returns (Node memory);
+
+    // === Queuing Stats ===
+
+    /// @notice Get the next available epoch for new registration
+    function nextRegisterationEpoch() external view returns (uint64);
+    /// @notice Get the number of pending registration requests in the waiting queue
+    function numPendingRegistrations() external view returns (uint64);
+    /// @notice Get the next available epoch for exit
+    function nextExitEpoch() external view returns (uint64);
+    /// @notice Get the number of pending exit requests in the waiting queue
+    function numPendingExit() external view returns (uint64);
+
+    // === Write APIs ===
+
+    /// @notice Register a validator in the stake table, transfer of tokens incurred!
+    ///
+    /// @param blsVK The BLS verification key
+    /// @param schnorrVK The Schnorr verification key (as the auxiliary info)
+    /// @param amount The amount to register
+    /// @param stakeType The type of staking (native or restaking)
+    /// @param blsSig The BLS signature that authenticates the `blsVK` field
+    /// @param maxWait The maximum epoch the sender is waiting to wait to be included (cannot be
+    /// smaller than the current epoch)
+    ///
+    /// @return success status
+    ///
+    /// @dev No validity check on `schnorrVK`, as it's assumed to be sender's responsibility,
+    /// the contract only treat it as auxiliary info submitted by `blsVK`.
+    function register(
+        BN254.G1Point calldata blsVK,
+        EdOnBN254.EdOnBN254Point calldata schnorrVK,
+        uint64 amount,
```

Comment:
> [chancharles92](https://github.com/chancharles92)
Is uint64 type sufficient for amount?

Reply:
I'm following our [doc here](https://docs.espressosys.com/sequencer/~/revisions/oijyFXG7VlZ4pxqo4Lgq/espresso-sequencer-architecture/internal-functionality/sequencer-contract/stake-table#data-structures) by @jbearer
 
64 bits should be enough?

cc @chancharles92

Reply:
64 bits is definitely enough *precision* for the contract. We may need to have some multiplier to convert from stake amounts to BEANs, since it's customary for ERC20s to have like 9 decimal places of precision, but we can definitely limit stake amounts to being a whole number or even a multiple of BEANs.

Reply:
> it's customary for ERC20s to have like 9 decimal places of precision

I'm not sure where "9" comes from, when reading openzepplin's [`IERC20` doc here](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#ERC20-decimals--), it says "18" is the default.
btw, I'm not suggesting we should go for 18, I'm fine with whatever decimals we find most sensible.

 > we can definitely limit stake amounts to being a whole number or even a multiple of BEANs.

with 9 decimal places, just absorbing the whole precision, the largest amount is still `2**55~= 3.6*10^16`, should be more than enough.

In short, we should be fine with `uint64` here.